### PR TITLE
Move StockLocation#create_stock_items to a background queue

### PR DESCRIPTION
**Pull request by sebastian-palma**
The idea is to make `StockLocation#create_stock_items` be executed as a background job, so I moved the method logic to a class that can be invoked in the job file.

I added the file spec, as the `create_stock_items` method is only being mentioned [here](https://github.com/spree/spree/blob/main/core/spec/models/spree/stock_location_spec.rb#L71), but it seems there are no tests checking the logic.
https://spark-solutions.atlassian.net/browse/SD-1398

### CODE REVIEW

File: core/app/services/spree/stock_locations/stock_items/create.rb [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+              Hash[
+                'stock_location_id', stock_location.id,
+                'variant_id', variant_id,
+                'backorderable', stock_location.backorderable_default,
+                'created_at', Time.current,
+                'updated_at', Time.current
+              ]
+            end
+            if prepared_stock_items.any?
+              stock_location.stock_items.insert_all(prepared_stock_items)
```
Comment (damianlegawiec) [inline core/app/services/spree/stock_locations/stock_items/create.rb:19]: @sebastian-palma let's also pass `unique_by` option to make sure we're skipping duplicate records, also let's add a test scenario with this

File: core/db/migrate/20210906065456_add_unique_stock_item_stock_location_variant_backord_index.rb [hunk 2]
```diff
@@ -0,0 +1,5 @@
+class AddUniqueStockItemStockLocationVariantBackordIndex < ActiveRecord::Migration[5.2]
+  def change
+    add_index :spree_stock_items, [:stock_location_id, :variant_id, :backorderable], name: 'stock_item_by_loc_var_id_and_backorderable', unique: true
```
Comment (damianlegawiec) [inline core/db/migrate/20210906065456_add_unique_stock_item_stock_location_variant_backord_index.rb:3]:
```suggestion
    add_index :spree_stock_items, [:stock_location_id, :variant_id, :deleted_at], name: 'stock_item_by_loc_var_id_and_backorderable', unique: true
```

based on

https://github.com/spree/spree/blob/82c6095a2fc3c2c81fbda989d951d964422d636b/core/app/models/spree/stock_item.rb#L12

Comment (sebastian-palma) [reply core/db/migrate/20210906065456_add_unique_stock_item_stock_location_variant_backord_index.rb:3]: Shouldn't this validation contain the `variant_id` as well?

Comment (damianlegawiec) [reply core/db/migrate/20210906065456_add_unique_stock_item_stock_location_variant_backord_index.rb:3]: @sebastian-palma it validates `variant_id` in `stock_location_id deleted_at` scope

File: core/spec/services/spree/stock_locations/stock_items/create_spec.rb [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        before do
+          # Delete all stock_location.stock_items to start counting from 0.
+          stock_location.stock_items.unscope(:where).delete_all
+        end
+
+        let(:time_current) { Time.local(1990) }
+
+        context 'with prepared stock items' do
+          context 'with duplicate stock items by stock_location_id and variant' do
+            it 'removes duplicate stock_items before inserting to avoid duplicate ones' do
```
Comment (damianlegawiec) [inline core/spec/services/spree/stock_locations/stock_items/create_spec.rb:42]: @sebastian-palma this description is kinda cryptic - does it remove the previous records or just skip the duplicates? :) auto-removing stuff makes my skin crawl 😂

Comment (sebastian-palma) [reply core/spec/services/spree/stock_locations/stock_items/create_spec.rb:42]: > does it remove the previous records or just skip the duplicates

It kind of skips them. It doesn't touch the db before that, that's certainly a confusing description.

I wrote it looking at [these](https://github.com/spree/spree/blob/a46b93ea1088db99b6f3b921e6f92af437971294/core/app/services/spree/stock_locations/stock_items/create.rb#L9:L17) lines. As `prepared_stock_items` doesn't include the `deleted_at` values, the index we just created isn't used for the insertion, so we have to remove the duplicates by hand and then insert them

Comment (damianlegawiec) [reply core/spec/services/spree/stock_locations/stock_items/create_spec.rb:42]: @sebastian-palma  Let's change that description - it's misleading ATM

**Review by damianlegawiec — APPROVED** [review]

**Pull request merged**